### PR TITLE
Swap deep link

### PR DESCRIPTION
**Pull request by Ran73Cohen**
#128 #186

### CODE REVIEW

**Comment by pingustar** [comment]
@Ran73Cohen I think query params are more suitable for the swap widget

**Comment by ashachaf** [comment]
- [ ] update CTA to indicate token not supported 
when a user selects a token on market and then switch to limit, please do the following:
1. IF the token exists on the token list (in this case the limit order token list) -> proceed normally
2. ELSE (token is not on the list) -> update the CTA:
- disable the CTA
- identify if the token is in the from or to
- update text: "To/From token not supported"

**Comment by ashachaf** [comment]
> * [x]  update CTA to indicate token not supported
>   when a user selects a token on market and then switch to limit, please do the following:
> 
> 1. IF the token exists on the token list (in this case the limit order token list) -> proceed normally
> 2. ELSE (token is not on the list) -> update the CTA:
> 
> * disable the CTA
> * identify if the token is in the from or to
> * update text: ""You Pay"/"You Receive" token is not supported"

i do not see this flow in the preview url.

**Comment by ashachaf** [comment]
- [ ] unsupported token flow
1. the text on the button is not correct as it is always saying "your Pay"
2. can we use the token symbol instead in the message? for example: "USDB is not supported"
image

- [ ] case of ETH as the "you receive" token:
1. from market to limit -> update it to WETH
2. from limit to market -> update it to ETH

- [ ] case of ETH as "you pay" token should not disable the CTA
image

File: src/elements/swapWidget/SwapWidget.tsx [hunk 1]
```diff
@@ -47,33 +76,53 @@ export const SwapWidget = ({ isLimit, setIsLimit }: SwapWidgetProps) => {
...
+    }${isLimit ? '&limit=' + isLimit : ''}`;
+
+    if (url !== window.location.search) history.push(url);
+  };
 
   const switchTokens = () => {
     if (toToken) {
-      setFromToken(toToken);
-      setToToken(fromToken);
+      const url = `?from=${toToken.address}${
```
Comment (zavelevsky) [inline src/elements/swapWidget/SwapWidget.tsx:120]: This is very specific application code - shouldn't take place inside a widget

File: src/services/observables/triggers.ts [hunk 2]
```diff
@@ -54,6 +54,7 @@ export const loadCommonData = (dispatch: any) => {
   });
 
   keeperDaoSub = keeperDaoTokens$.subscribe((keeperDaoTokens) => {
+    dispatch(setKeeperDaoTokens(keeperDaoTokens));
```
Comment (zavelevsky) [inline src/services/observables/triggers.ts:57]: duplicate call to `setKeeperDaoTokens`

**Review by zavelevsky — CHANGES_REQUESTED** [review]

File: src/elements/earn/pools/poolsTable/PoolsTableCellActions.tsx [hunk 3]
```diff
@@ -3,12 +3,19 @@ import { Tooltip } from 'components/tooltip/Tooltip';
...
+import { EthNetworks } from 'services/web3/types';
 
 export const PoolsTableCellActions = (pool: Pool) => {
+  const { chainId } = useWeb3React();
   const href = pool.isProtected
     ? `https://app.bancor.network/eth/portfolio/stake/add/single/${pool.pool_dlt_id}`
     : `https://app.bancor.network/eth/pool/add/${pool.pool_dlt_id}`;
-
+  const tknAddress = pool.reserves.find(
+    (x) => x.address !== bntToken(chainId ? chainId : EthNetworks.Mainnet)
```
[the remark this answers is not in the document]
Comment (Ran73Cohen) [reply src/elements/earn/pools/poolsTable/PoolsTableCellActions.tsx:16]: Its just two reserves but yeah will place it outside

File: src/elements/earn/pools/poolsTable/PoolsTableCellActions.tsx [hunk 4]
```diff
@@ -3,12 +3,19 @@ import { Tooltip } from 'components/tooltip/Tooltip';
...
+  )?.address;
+  console.log('tknAddress', tknAddress);
```
[the remark this answers is not in the document]
Comment (Ran73Cohen) [reply src/elements/earn/pools/poolsTable/PoolsTableCellActions.tsx:18]: Thanks! forgot to

File: src/elements/earn/pools/poolsTable/PoolsTableCellActions.tsx [hunk 5]
```diff
@@ -27,7 +34,7 @@ export const PoolsTableCellActions = (pool: Pool) => {
         />
       </a>
       <NavLink
-        to="/"
+        to={tknAddress ? `/?from=${tknAddress}` : '/'}
```
[the remark this answers is not in the document]
Comment (Ran73Cohen) [reply src/elements/earn/pools/poolsTable/PoolsTableCellActions.tsx:37]: End slashes are removed from route by default, I can link to an empty route but when redirecting to root it is general practice to use /

File: src/elements/swapLimit/SwapLimit.tsx [hunk 6]
```diff
@@ -314,6 +316,21 @@ export const SwapLimit = ({
     handleFieldChanged(Field.rate, fromAmount, toAmount, val);
   };
 
+  const swapButtonText = () => {
+    if (!toToken) return 'Select a token';
+    else if (
+      keeperDaoTokens.findIndex((x) => x.address === fromToken.address) === -1
+    )
+      return 'You Pay token is not supported';
```
[the remark this answers is not in the document]
Comment (Ran73Cohen) [reply src/elements/swapLimit/SwapLimit.tsx]: Its like "You pay" as In the field but we changed it to the symbol itself

File: src/elements/swapLimit/SwapLimit.tsx [hunk 7]
```diff
@@ -295,6 +290,13 @@ export const SwapLimit = ({
       return true;
     if (!toToken) return true;
     if (!account) return false;
+    if (
```
[the remark this answers is not in the document]
Comment (Ran73Cohen) [reply src/elements/swapLimit/SwapLimit.tsx:293]: I agree, you can open an issue about swap limit and swap market

File: src/elements/swapWidget/SwapWidget.tsx [hunk 8]
```diff
@@ -47,33 +76,53 @@ export const SwapWidget = ({ isLimit, setIsLimit }: SwapWidgetProps) => {
...
-    ) {
-      const foundTo = findSetToken(toToken);
-      foundTo ? setToToken(foundTo) : setToToken(tokens[1]);
-    }
-  }, [tokens, fromToken, toToken]);
+    const url = `?from=${fromToken?.address}${
+      toToken ? '&to=' + toAddress : ''
+    }${limit ? '&limit=' + limit : ''}`;
+
+    if (!fromToken) history.push(limit ? '?limit=' + limit : '');
```
[the remark this answers is not in the document]
Comment (Ran73Cohen) [reply src/elements/swapWidget/SwapWidget.tsx:91]: Well thats the point of this task to create the deep link based on the tokens the user selects

File: src/services/observables/triggers.ts (hunk 2 above)
Comment (Ran73Cohen) [reply src/services/observables/triggers.ts:57]: stale

**Comment by ashachaf** [comment]
minor fix:
- [ ] switching between market to limit - carry the user input value
flow:
- open market
- select from/to tokens
- input value in the from token field
- switch to limit view
expected: keep the input value when switching to the limit view. 

- [ ] CTA text to connect a wallet
on market, we have a logical flow that shows "connect a wallet" text on the CTA if the user did not connect a wallet.
however, on limit, the text is "trade" and doesnt follow the same logic.
expected:
- limit and market should follow the same CTA text logic when user wallet not connected, tokens are not selected and no input amount.

**Pull request merged**